### ISSUE TITLE
Bought Price Postmortem [July 13, 2018]

### ISSUE THREAD

**Issue by PTDefenderDev** [PTDefender/Welcome#32]
My goal for this write up is to clearly and honestly explain exactly what's happening with the incorrect bought prices in ProfitTrailer, which in turn can cause coins to be sold at unexpected times.

This write up will focus on these main points:

1. How ProfitTrailer tracks bought price
2. When ProfitTrailer loses the bought price
3. How PTDefender uses bought price
4. Why incorrect bought prices were set in ProfitTrailer
5. How we've mitigated the issue

## How ProfitTrailer tracks bought prices

**Note:** I must start by saying that ProfitTrailer does not clearly document how they track and reset bought price. Everything I am about to explain is from our own extensive testing and what we have observed.

When you first buy a coin, your bought price will become the price that the first order executed at. As you buy more coins the bought price is adjusted using a basic weighted average formula where the amount you bought is the weight. This is clearly (although we think incorrectly, we could never calculate an exact match) explained in the ProfitTrailer wiki here: [https://wiki.profittrailer.com/doku.php?id=hotconfig#bought_price](https://wiki.profittrailer.com/doku.php?id=hotconfig#bought_price)

As we were building PTDefender it was critical to understand how the bought price changes over time, it's ultimately how we know how much money you've recovered and when we can stop defending. What we observed in ProfitTrailer v2.0.13 was that as we sold coins the bought price would change. This caused us to incorrectly think that ProfitTrailer was not using a simple weighted average calculation but something else like a First-in-first-out (FIFO) calculation. As we did more testing we incorrectly concluded that it was a FIFO calculation. More on this later.

The biggest problem with ProfitTrailer and bought prices is that it often loses the prices and displays 0 in the dashboard. When this happens users are instructed to use the Hotconfig to either set the bought price if they know it, or try resetting the stored average. This is especially problematic for PTDefender as ProfitTrailer loses the bought price as soon as we place a pending sell order or cancel a pending sell order. Because of this we are forced to use the Hotconfig API to set bought price so ProfitTrailer can continue working on your bag. More on this later.

When resetting the stored average we observed that ProfitTrailer is no longer using the FIFO calculation that it used before the price was lost. Instead it used a simple weighted average calculation using just buys since the last time you had 0 balance for that coin. More on this later.

## When ProfitTrailer loses the bought price

ProfitTrailer has a difficult time maintaining the bought price when external trades occur. Specifically ProfitTrailer will lose the bought price in all of these cases (as of v2.0.14):

1. You place a pending sell order on the exchange
2. You place and fill a sell order for an amount not divisible by 10
3. You cancel a pending sell order

## How PTDefender uses bought price

In an ideal world we would never use the bought price and simply look at the profit percentage that ProfitTrailer reports. With that percentage we would know when to stop defending once it meets your target.

The reason we can't rely on this is because as soon as you start defending we place a pending sell order on the exchange to begin breaking up your bag. As soon as we do this ProfitTrailer will lose the bought price for that coin and will never rebuy it. Because of this we must reset the bought price to the previous value, this works as expected.

Where we start to run into issues is:

1. When ProfitTrailer loses the bought price after PTDefender buys
2. When ProfitTrailer loses the bought price after a defended coin is canceled due to insufficient balance

In both of these cases the bought price is no longer the previous value that we had stored. In order to properly continue defending the coin we need a way to calculate the new bought price and set it using the Hotconfig API.

## Why incorrect bought prices were set in ProfitTrailer

During the first few days after we released PTDefender a report came in that PTDefender had caused ProfitTrailer to sell their coins at a loss. At this point we had not implemented any custom calculations for setting bought price. What we did do, however, was ask ProfitTrailer to reset the stored average in the following cases:

1. When a user stops defending
2. When a defended coin is canceled
3. After we re-buy

After extensive testing we found that in certain cases ProfitTrailer would use a different value when resetting the stored average than it previously calculated. Part of this may be the frequency in which we need to reset the stored average (I don't think any other add-ons have attempted something like this yet) and we also think it has something to with the pending order that is placed on the exchange, and ProfitTrailer having the incorrect balance of coins.

Seeing this lead us to implement our own calculations for average price which we could use as frequently as we need to, and they would consistently set the price using the same math. If you remember above, we incorrectly thought that ProfitTrailer was using a FIFO calculation for bought price. We updated the app around 1.3.0 to start using this new FIFO calculation. In our testing at the time the calculations were matching up really well with what ProfitTrailer was reporting. 

Soon after, ProfitTrailer released version 2.0.14 which had a handful of fixes for average price. We failed to retest our calculations using 2.0.14 until 2 days ago when we started getting an influx of reports that PTDefender was causing coins to be sold with an incorrect bought price. After extensively testing 2.0.14 (and then re-testing 2.0.13 with a handful of new cases) we realized that ProfitTrailer is using a simple weighted average calculation and not updating the average price when sell orders are filled. The previous behavior we saw where the average price was being updated after sells was most likely a bug.

## How we've mitigated the issue

In order to mitigate the issue we needed to take a couple important steps:

1. Prevent people from defending new coins on versions older than 1.7.0 (just released)
2. Update our bought price calculations to use a simple weighted average that matches ProfitTrailer
3. Set the bought price to our calculation when you start defending a coin, this ensures the bought price is always consistent throughout the entire process
4. Provide a table in the start defending modal that shows you exactly what will be used for your bought price

As previously mentioned, even with the simple weighted average calculation, we haven't been able to calculate bought prices that 100% match what ProfitTrailer is showing. In all of our testing the bought prices end up being off by 0.1% - 0.2%. This is why think it's really important to introduce the table in the start defending screen that shows exactly what we are calculating. To account for exchange fees we always return a bought price that is 0.2% higher than what we calculate. This also allows us to er on the side of caution and not set a bought price that is too low and could potentially trigger a sell.

Release Notes for v1.7.0: https://github.com/PTDefender/Welcome/releases/tag/1.7.0

**Comment by djbadders** [comment]
Thanks for the great explanation. Will try 1.7 later and feed back the results.

**Comment by djbadders** [comment]
Ok using 1.7.0 I have had a weird issue with OAX. I cancelled all defending after upgrading and then starting defending 6 coins including OAX. All looked good. OAX was bought by PT but the price got set wrong with it ending up at around -22%. Looking at the logs I can see two average prices being set after the buy, one correct one and one erroneous one. Logs below:

~~~~
2018-07-14 15:43:39 INFO AverageCalculator - OAX using bought price 0.0000827329
2018-07-14 15:44:05 INFO AverageCalculator - ADA using bought price 0.0000362982
2018-07-14 15:44:18 INFO Util - Detected configuration changes
2018-07-14 15:45:24 INFO AverageCalculator - AION using bought price 0.0001457528
2018-07-14 15:45:37 INFO Util - Detected configuration changes
2018-07-14 15:46:15 INFO Util - Detected configuration changes
2018-07-14 15:46:46 INFO ProfitTrailerService - Store last order before buy/sell orderid: 3231748
2018-07-14 15:46:47 INFO ProfitTrailerService - Sell order for AIONBTC sold amount 173.310000 for price 0.000148 -- {"symbol":"AIONBTC","orderId":19361343,"clientOrderId":"FLi63hshQwTNKiz54tZ9wU","transactTime":1531583207242,"price":"0.00014810","origQty":"173.31000000","executedQty":"173.31000000","status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"SELL"}
2018-07-14 15:46:52 INFO ProfitTrailerService - Get order information -- {"symbol":"AIONBTC","orderId":19361343,"clientOrderId":"FLi63hshQwTNKiz54tZ9wU","price":0.00014810,"origQty":173.31000000,"executedQty":173.31000000,"status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"SELL","stopPrice":0E-8,"icebergQty":0E-8,"time":1531583207242}
2018-07-14 15:46:53 INFO ProfitTrailerService - Check last order after buy/sell orderid:3250547, retries 0
2018-07-14 15:46:53 INFO StrategyRunnerHelper - SELLLOG: {"buyProfit":0.0,"lastOrderNumber":"","unfilledAmount":0.0,"market":"AIONBTC","profit":1.480960548885091,"orderbookProfit":1.4123499142367104,"averageCalculator":{"totalCost":0.025260417768000000,"totalAmount":173.31000000,"totalAmountWithSold":173.31000000,"avgPrice":0.0001457528,"avgCost":0.025260417768000000,"firstBoughtDate":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":45,"second":24,"nano":849000000}},"totalWeightedPrice":0.025260417768000000,"fee":0.2,"usedBuyHistory":[{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":45,"second":24,"nano":849000000}},"rate":"0.0001457528","amount":"173.31000000","total":"0.025260417768000000","fee":"0","orderNumber":"","type":"buy","category":""}],"usedSellHistory":[]},"currentPrice":1.482E-4,"buyStrategies":[],"sellStrategies":[{"type":"SELL","name":"GAIN","entryValue":0.1,"entryValueLimit":0.0,"triggerValue":0.0,"currentValue":1.7554030874785689,"currentValuePercentage":0.0,"decimals":2,"positive":"true trailing..."}],"volume":86.09878206,"percChange":0.02407,"boughtTimes":0,"originalTrailingData":{"originalTrigger":0.0,"originalWidth":0.0,"originalValue":0.1,"firstTrailTime":{"date":{"year":2018,"month":7,"day":12},"time":{"hour":2,"minute":7,"second":27,"nano":271000000}},"lastTrailTime":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":47,"second":46,"nano":603000000}}}}
2018-07-14 15:47:13 INFO ProfitTrailerService - Cache Heartbeat
2018-07-14 15:51:10 INFO ProfitTrailerService - PAIRS Heartbeat
2018-07-14 15:51:18 INFO Util - Detected configuration changes
2018-07-14 15:51:36 INFO ProfitTrailerService - DCA Heartbeat
2018-07-14 15:53:22 INFO AverageCalculator - Please set BNB_bought_price
2018-07-14 15:53:22 INFO AverageCalculator - BALANCE:{"available":2.86676418,"originalAvailable":2.86676418,"onOrders":0E-8,"btcValue":0.005549768776062,"dust":false} -- [{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":4},"time":{"hour":20,"minute":0,"second":46,"nano":352000000}},"rate":"0.0021","amount":"3.47642577","total":"0.007300494117","fee":"0","orderNumber":"17191561","type":"buy","category":""}]
2018-07-14 15:56:19 INFO Util - Detected configuration changes
2018-07-14 15:57:17 INFO ProfitTrailerService - Cache Heartbeat
2018-07-14 16:01:14 INFO ProfitTrailerService - PAIRS Heartbeat
2018-07-14 16:01:19 INFO Util - Detected configuration changes
2018-07-14 16:01:36 INFO ProfitTrailerService - Store last order before buy/sell orderid: 1282978
2018-07-14 16:01:36 INFO ProfitTrailerService - Buy order for OAXBTC -- {"symbol":"OAXBTC","orderId":11810422,"clientOrderId":"a341zxvejubVgKRjqncsib","transactTime":1531584096432,"price":"0.00005151","origQty":"200.00000000","executedQty":"200.00000000","status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"BUY"}
2018-07-14 16:01:42 INFO ProfitTrailerService - Get order information -- {"symbol":"OAXBTC","orderId":11810422,"clientOrderId":"a341zxvejubVgKRjqncsib","price":0.00005151,"origQty":200.00000000,"executedQty":200.00000000,"status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"BUY","stopPrice":0E-8,"icebergQty":0E-8,"time":1531584096432}
2018-07-14 16:01:42 INFO ProfitTrailerService - Check last order after buy/sell orderid:1287099, retries 0
2018-07-14 16:01:42 INFO NormalStrategyRunner - Normal Pair - currency pair OAXBTC, MTP 35, TTP 29
2018-07-14 16:01:42 INFO NormalStrategyRunner - BUYLOG: {"market":"OAXBTC","profit":0.0,"orderbookProfit":0.0,"currentPrice":5.151E-5,"buyStrategies":[{"type":"BUY","name":"EMAGAIN","entryValue":-0.22,"entryValueLimit":0.0,"triggerValue":0.0,"currentValue":0.11186942387326759,"currentValuePercentage":0.0,"decimals":2,"positive":"true trailing..."}],"sellStrategies":[],"volume":15.51167197,"percChange":0.00807,"boughtTimes":0,"originalTrailingData":{"originalTrigger":0.0,"originalWidth":0.0,"originalValue":5.163E-5,"firstTrailTime":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":43,"second":30,"nano":674000000}},"lastTrailTime":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":3,"second":35,"nano":998000000}}}}
2018-07-14 16:01:48 INFO ProfitTrailerService - DCA Heartbeat
2018-07-14 16:01:48 INFO AverageCalculator - Please set OAX_bought_price
2018-07-14 16:01:48 INFO AverageCalculator - BALANCE:{"available":203.00000000,"originalAvailable":203.00000000,"onOrders":635.00000000,"btcValue":0.01042202,"dust":false} -- [{"globalTradeID":0,"tradeID":"1287100","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":1,"second":47,"nano":614000000}},"rate":"0.00005151","amount":"119.0","total":"0.00612969","fee":"0.0","orderNumber":"1287100","type":"buy","category":""},{"globalTradeID":0,"tradeID":"1287099","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":1,"second":36,"nano":432000000}},"rate":"0.00005151","amount":"200.0","total":"0.010302","fee":"0.0","orderNumber":"1287099","type":"buy","category":""},{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":43,"second":39,"nano":577000000}},"rate":"0.0000827329","amount":"3.00000000","total":"0.000248198700000000","fee":"0","orderNumber":"","type":"buy","category":""}]
2018-07-14 16:01:48 INFO AverageCalculator - NEW BUY CALCULATED: {"totalCost":0,"totalAmount":203.00000000,"totalAmountWithSold":0,"avgPrice":0,"avgCost":0,"totalWeightedPrice":0,"fee":0.2,"usedBuyHistory":[{"globalTradeID":0,"tradeID":"1287100","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":1,"second":47,"nano":614000000}},"rate":"0.00005151","amount":"119.0","total":"0.00612969","fee":"0.0","orderNumber":"1287100","type":"buy","category":""},{"globalTradeID":0,"tradeID":"1287099","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":1,"second":36,"nano":432000000}},"rate":"0.00005151","amount":"200.0","total":"0.010302","fee":"0.0","orderNumber":"1287099","type":"buy","category":""},{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":15,"minute":43,"second":39,"nano":577000000}},"rate":"0.0000827329","amount":"3.00000000","total":"0.000248198700000000","fee":"0","orderNumber":"","type":"buy","category":""}],"usedSellHistory":[]}
2018-07-14 16:01:59 INFO AverageCalculator - OAX using bought price 0.0000516130
2018-07-14 16:02:31 INFO AverageCalculator - OAX using bought price 0.0000668847
~~~~

Trade history:
![image](https://user-images.githubusercontent.com/34887832/42726707-d96da7c8-8790-11e8-8f70-49d539f7fb29.png)

**Comment by djbadders** [comment]
It has has miscalculated the average price again after another purchase of OAX, again set to around -22% the logs are:
~~~~
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:31 INFO ProfitTrailerService - Buy order for OAXBTC -- {"symbol":"OAXBTC","orderId":11813377,"clientOrderId":"xH9tUlObRaflN86Pc0gh4o","transactTime":1531587870967,"price":"0.00005033","origQty":"209.00000000","executedQty":"209.00000000","status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"BUY"}
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:36 INFO ProfitTrailerService - Get order information -- {"symbol":"OAXBTC","orderId":11813377,"clientOrderId":"xH9tUlObRaflN86Pc0gh4o","price":0.00005033,"origQty":209.00000000,"executedQty":209.00000000,"status":"FILLED","timeInForce":"IOC","type":"LIMIT","side":"BUY","stopPrice":0E-8,"icebergQty":0E-8,"time":1531587870967}
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:36 INFO ProfitTrailerService - Check last order after buy/sell orderid:1287373, retries 0
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:36 INFO NormalStrategyRunner - Normal Pair - currency pair OAXBTC, MTP 35, TTP 29
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:36 INFO NormalStrategyRunner - BUYLOG: {"market":"OAXBTC","profit":0.0,"orderbookProfit":0.0,"currentPrice":5.033E-5,"buyStrategies":[{"type":"BUY","name":"EMAGAIN","entryValue":-0.22,"entryValueLimit":0.0,"triggerValue":0.0,"currentValue":-1.928859155490814,"currentValuePercentage":0.0,"decimals":2,"positive":"true trailing..."}],"sellStrategies":[],"volume":19.53699672,"percChange":-0.01233,"boughtTimes":0,"originalTrailingData":{"originalTrigger":0.0,"originalWidth":0.0,"originalValue":5.145E-5,"firstTrailTime":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":1,"second":50,"nano":459000000}},"lastTrailTime":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":6,"second":30,"nano":576000000}}}}
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:43 INFO AverageCalculator - Please set OAX_bought_price
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:43 INFO AverageCalculator - BALANCE:{"available":212.00000000,"originalAvailable":212.00000000,"onOrders":632.00000000,"btcValue":0.01063392,"dust":false} -- [{"globalTradeID":0,"tradeID":"1287373","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":4,"second":30,"nano":967000000}},"rate":"0.00005033","amount":"91.0","total":"0.00458003","fee":"0.0","orderNumber":"1287373","type":"buy","category":""},{"globalTradeID":0,"tradeID":"1287372","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":4,"second":30,"nano":967000000}},"rate":"0.00005032","amount":"118.0","total":"0.00593776","fee":"0.0","orderNumber":"1287372","type":"buy","category":""},{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":49,"second":30,"nano":229000000}},"rate":"0.0000767527","amount":"635.00000000","total":"0.048737964500000000","fee":"0","orderNumber":"","type":"buy","category":""}]
0|BTC2 Pro | 2018-07-14 17:04 +00:00: 2018-07-14 17:04:43 INFO AverageCalculator - NEW BUY CALCULATED: {"totalCost":0,"totalAmount":212.00000000,"totalAmountWithSold":0,"avgPrice":0,"avgCost":0,"totalWeightedPrice":0,"fee":0.2,"usedBuyHistory":[{"globalTradeID":0,"tradeID":"1287373","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":4,"second":30,"nano":967000000}},"rate":"0.00005033","amount":"91.0","total":"0.00458003","fee":"0.0","orderNumber":"1287373","type":"buy","category":""},{"globalTradeID":0,"tradeID":"1287372","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":17,"minute":4,"second":30,"nano":967000000}},"rate":"0.00005032","amount":"118.0","total":"0.00593776","fee":"0.0","orderNumber":"1287372","type":"buy","category":""},{"globalTradeID":0,"tradeID":"SETBOUGHTPRICE","date":{"date":{"year":2018,"month":7,"day":14},"time":{"hour":16,"minute":49,"second":30,"nano":229000000}},"rate":"0.0000767527","amount":"635.00000000","total":"0.048737964500000000","fee":"0","orderNumber":"","type":"buy","category":""}],"usedSellHistory":[]}
0|BTC2 Pro | 2018-07-14 17:05 +00:00: 2018-07-14 17:05:16 INFO AverageCalculator - AION using bought price 0.0001486968
0|BTC2 Pro | 2018-07-14 17:05 +00:00: 2018-07-14 17:05:36 INFO AverageCalculator - OAX using bought price 0.0000636179
0|BTC2 Pro | 2018-07-14 17:05 +00:00: 2018-07-14 17:05:48 INFO AverageCalculator - SNGLS using bought price 0.0000065685
0|BTC2 Pro | 2018-07-14 17:05 +00:00: 2018-07-14 17:05:51 INFO AverageCalculator - AION using bought price 0.0004551
0|BTC2 Pro | 2018-07-14 17:06 +00:00: 2018-07-14 17:06:18 INFO Util - Detected configuration changes
0|BTC2 Pro | 2018-07-14 17:07 +00:00: 2018-07-14 17:07:31 INFO ProfitTrailerService - Cache Heartbeat
0|BTC2 Pro | 2018-07-14 17:11 +00:00: 2018-07-14 17:11:17 INFO Util - Detected configuration changes
0|BTC2 Pro | 2018-07-14 17:11 +00:00: 2018-07-14 17:11:42 INFO ProfitTrailerService - PAIRS Heartbeat
0|BTC2 Pro | 2018-07-14 17:12 +00:00: 2018-07-14 17:12:12 INFO ProfitTrailerService - DCA Heartbeat
~~~~

**Comment by djbadders** [comment]
Also in the UI it is showing 0 for recovered value, even though it is round 3
![image](https://user-images.githubusercontent.com/34887832/42726806-3e33256a-8792-11e8-87e8-d0bb5af23156.png)

![image](https://user-images.githubusercontent.com/34887832/42726809-4f5cf604-8792-11e8-931b-022a3261d716.png)

**Comment by djbadders** [comment]
I am going to stop PTD again as this is not right at all :(

**Comment by djbadders** [comment]
@PTDefenderDev ^^^

**Comment by djbadders** [comment]
I can confirm I am seeing miscalculations after buys across both 1.7.0 instances and multiple pairs. Each time the % is around -22% for the pair.

**Comment by PTDefenderDev** [comment]
@djbadders Looking over all this now. Can you send me screens shots of what shows up in the "Start Defending" modal? Specifically that table that has our calculations before you start defending

**Comment by djbadders** [comment]
![image](https://user-images.githubusercontent.com/34887832/42727852-b9766b52-87a5-11e8-98da-403347ea979d.png)

![image](https://user-images.githubusercontent.com/34887832/42727861-d1b1dfbc-87a5-11e8-9d68-9788154b17c9.png)

**Comment by PTDefenderDev** [comment]
@djbadders In that screen, does an avg price of 0.00007253 seem correct to you?

**Comment by djbadders** [comment]
Yes it does

**Comment by PTDefenderDev** [comment]
Just released 1.7.1 that with the override settings and important bug fixes that could cause out hotconfig to not get set: https://github.com/PTDefender/Welcome/releases/tag/1.7.1

**Comment by djbadders** [comment]
Seems to be behaving itself today 👍 Will keep an eye on it, but I like the changes you made.
